File: NeetCode/04_sliding_window/01_longest_string_without_repeating_characters/longest_string_without_repeating_characters.cpp
#include <string>
#include <unordered_set>
#include <cmath>

using namespace std;

class Solution
{
  public:
    int lengthOfLongestSubstring(string s)
    {
        size_t              first = 0;
        size_t              last = 0;
        int                 longest = 0;
        unordered_set<char> char_track;
        while (last < s.length())
        {
            // Locate and skip to duplicate letter
            while (char_track.contains(s[last]))
            {
                char_track.erase(s[first]);
                ++first;
            }
            char_track.insert(s[last]);
            ++last;
            longest = max(static_cast<int>(last - first), longest);
        }
        return longest;
    }
};

#include <iostream>
#include <assert.h>

int main()
{
    Solution sln;
    string   test {"pwwkew"};
    test = "aab";
    assert(sln.lengthOfLongestSubstring(test) == 2);

    test = "nfpdmpi";
    assert(sln.lengthOfLongestSubstring(test) == 5);

    test = "abc";
    assert(sln.lengthOfLongestSubstring(test) == 3);

    test = "tmmzuxt";
    assert(sln.lengthOfLongestSubstring(test) == 5);
}

File: NeetCode/04_sliding_window/01_longest_string_without_repeating_characters/longest_string_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        begin: int = 0
        end: int = 0
        char_track = set()
        longest: int = 0
        while end < len(s):
            # Skip all letters until duplicate is erased as well
            while s[end] in char_track:
                char_track.remove(s[begin])
                begin += 1
            char_track.add(s[end])
            end += 1
            longest = max(longest, end - begin)
        return longest
